File: apps/web/lib/files.ts
/**
 * Client-side files client for the read-only Files API (T0308 UI over the
 * T0307 surface).
 *
 * Wraps the four JSON reads the Files page renders —
 *   GET /api/v1/projects/{id}/files/tree?ref=&path=
 *   GET /api/v1/projects/{id}/files/content?ref=&path=
 *   GET /api/v1/projects/{id}/files/history?ref=&path=&limit=
 * — and builds the two stream URLs the page links to (never fetches):
 *   GET /api/v1/projects/{id}/files/raw?ref=&path=   (the download link)
 *   GET /api/v1/projects/{id}/files/diff?sha=        (the raw diff link)
 *
 * The client is read-only by construction: it has no method that sends
 * anything but GET, and the two URL builders exist precisely so the page
 * renders plain <a href> links instead of mutating calls. All calls carry
 * the session cookie (credentials: include) — anonymous callers read
 * public projects only, exactly like the projects client.
 *
 * The module is import-free by construction (like lib/auth.ts and
 * lib/projects.ts), so the node:test suite (files.test.mjs) runs it
 * through Node's type stripping. The ApiError shape mirrors lib/projects.ts.
 */

/** The API error envelope (docs/22 §5): stable codes, no stack traces. */
export interface ErrorEnvelope {
  code: string;
  message: string;
  request_id?: string;
  retryable?: boolean;
}

/**
 * ApiError carries the API's envelope: the UI renders `message` for the
 * codes it knows and a generic line otherwise — never raw transport errors.
 * (Same shape as lib/auth.ts / lib/projects.ts ApiError.)
 */
export class ApiError extends Error {
  readonly code: string;
  readonly status: number;
  readonly retryable: boolean;

  constructor(status: number, envelope: ErrorEnvelope) {
    super(envelope.message);
    this.name = "ApiError";
    this.code = envelope.code;
    this.status = status;
    this.retryable = envelope.retryable ?? status === 503;
  }
}

/** The wire tree-entry shape (cmd/api/fileshttp treeEntryPayload). */
export interface FilesTreeEntry {
  name: string;
  path: string;
  /** "blob", "tree", "symlink" or "gitlink" — derived from the git mode. */
  type: string;
  /** The raw git mode ("100644", "040000", …). */
  mode: string;
  size: number;
  sha: string;
}

/** The wire tree-listing shape (cmd/api/fileshttp treePayload). */
export interface FilesTreeListing {
  ref: string;
  path: string;
  sha: string;
  entries: FilesTreeEntry[];
}

/** The displayed form of a blob-pointer manifest (docs/17 §2). */
export interface FilesBlobPointer {
  content_hash: string;
  size_bytes: number;
  blob_id?: string;
}

/** The wire file-view shape (cmd/api/fileshttp filePayload). */
export interface FilesFileView {
  name: string;
  path: string;
  type: string;
  mode: string;
  size: number;
  sha: string;
  /** "text", "binary", "too_large", "symlink" or "gitlink". */
  kind: string;
  content: string;
  truncated: boolean;
  target?: string;
  blob_pointer?: FilesBlobPointer;
}

/** The wire commit shape (cmd/api/fileshttp commitPayload). */
export interface FilesCommitEntry {
  sha: string;
  author: string;
  author_email: string;
  date: string;
  message: string;
}

/** Options shared by the JSON reads. */
export interface FilesReadOptions {
  /** Branch name or full commit SHA; "" = the canonical main branch. */
  ref?: string;
  path?: string;
}

/** The files client: one per API origin. */
export interface FilesClient {
  /** One directory listing at ref/path ("" = root of main). */
  tree(projectId: string, opts: FilesReadOptions): Promise<FilesTreeListing>;
  /** One file's safe preview (the content endpoint reads blobs only). */
  content(projectId: string, opts: FilesReadOptions): Promise<FilesFileView>;
  /** The commit history touching path at ref, newest first. */
  history(projectId: string, opts: FilesReadOptions & { limit?: number }): Promise<FilesCommitEntry[]>;
  /** The download link for one file (the raw channel — plain <a href>). */
  rawUrl(projectId: string, opts: { ref?: string; path: string }): string;
  /** The raw diff link for one commit (the patch channel — plain <a href>). */
  diffUrl(projectId: string, sha: string): string;
}

type FetchLike = (
  input: string,
  init?: { method?: string; headers?: Record<string, string>; credentials?: string },
) => Promise<{ status: number; json(): Promise<unknown> }>;

/** Options for createFilesClient. */
export interface FilesClientOptions {
  /** Fetch implementation (tests inject a fake); default: global fetch. */
  fetch?: FetchLike;
}

/** Build the files client for one API origin. */
export function createFilesClient(
  apiBaseUrl: string,
  options: FilesClientOptions = {},
): FilesClient {
  const fetchFn: FetchLike = options.fetch ?? (fetch as unknown as FetchLike);

  /** One GET against the API; the session cookie rides along. */
  async function call(path: string): Promise<{ status: number; body: unknown }> {
    const res = await fetchFn(apiBaseUrl + path, {
      method: "GET",
      credentials: "include",
    });
    let parsed: unknown = null;
    try {
      parsed = await res.json();
    } catch {
      parsed = null; // non-JSON bodies (e.g. a proxy error page)
    }
    return { status: res.status, body: parsed };
  }

  /** Throw ApiError for any non-2xx, decoding the stable envelope. */
  function fail(status: number, body: unknown): never {
    if (typeof body === "object" && body !== null && "code" in body) {
      throw new ApiError(status, body as ErrorEnvelope);
    }
    throw new ApiError(status, {
      code: "UNKNOWN",
      message: "the server answered unexpectedly; try again",
    });
  }

  /** One files route path with the ref/path query, never empty-ref dirty. */
  function route(projectId: string, endpoint: string, opts: FilesReadOptions): string {
    const params = new URLSearchParams();
    if (opts.ref !== undefined && opts.ref !== "") params.set("ref", opts.ref);
    if (opts.path !== undefined && opts.path !== "") params.set("path", opts.path);
    const q = params.toString();
    return (
      `/api/v1/projects/${encodeURIComponent(projectId)}/files/${endpoint}` +
      (q === "" ? "" : `?${q}`)
    );
  }

  function asListing(body: unknown): FilesTreeListing {
    if (
      typeof body !== "object" || body === null ||
      !("entries" in body) || !Array.isArray(body.entries)
    ) {
      throw new Error("unexpected tree response shape");
    }
    return body as FilesTreeListing;
  }

  function asFileView(body: unknown): FilesFileView {
    if (
      typeof body !== "object" || body === null ||
      !("path" in body) || typeof body.path !== "string" ||
      !("kind" in body) || typeof body.kind !== "string"
    ) {
      throw new Error("unexpected file content response shape");
    }
    return body as FilesFileView;
  }

  return {
    async tree(projectId, opts) {
      const result = await call(route(projectId, "tree", opts));
      if (result.status >= 400) fail(result.status, result.body);
      return asListing(result.body);
    },
    async content(projectId, opts) {
      const result = await call(route(projectId, "content", opts));
      if (result.status >= 400) fail(result.status, result.body);
      return asFileView(result.body);
    },
    async history(projectId, opts) {
      const params = new URLSearchParams();
      if (opts.ref !== undefined && opts.ref !== "") params.set("ref", opts.ref);
      if (opts.path !== undefined && opts.path !== "") params.set("path", opts.path);
      if (opts.limit !== undefined) params.set("limit", String(opts.limit));
      const q = params.toString();
      const result = await call(
        `/api/v1/projects/${encodeURIComponent(projectId)}/files/history` +
          (q === "" ? "" : `?${q}`),
      );
      if (result.status >= 400) fail(result.status, result.body);
      const body = result.body;
      if (
        typeof body !== "object" || body === null ||
        !("entries" in body) || !Array.isArray(body.entries)
      ) {
        throw new Error("unexpected history response shape");
      }
      return body.entries as FilesCommitEntry[];
    },
    rawUrl(projectId, opts) {
      // Absolute: the API is another origin than the page — the link
      // target must not resolve against the web origin.
      return apiBaseUrl + route(projectId, "raw", opts);
    },
    diffUrl(projectId, sha) {
      return (
        apiBaseUrl +
        `/api/v1/projects/${encodeURIComponent(projectId)}/files/diff` +
        `?sha=${encodeURIComponent(sha)}`
      );
    },
  };
}

/** Human-facing message for a stable files API code. */
export function messageForFileCode(code: string): string {
  switch (code) {
    case "PROJECT_NOT_FOUND":
      return "This project does not exist, or you do not have access to it.";
    case "FILES_DISABLED":
      return "File browsing is not configured for this server.";
    case "PROJECT_NOT_PROVISIONED":
      return "This project's repository is still being provisioned. Try again shortly.";
    case "FILES_INVALID_REF":
      return "That branch or commit does not look valid.";
    case "FILES_INVALID_SHA":
      return "That commit does not look valid.";
    case "FILES_INVALID_PATH":
      return "That path does not look valid.";
    case "FILES_NOT_FOUND":
      return "No such file, directory or commit in this repository.";
    case "FILES_PATH_IS_DIRECTORY":
      return "That path is a directory — pick a file to preview it.";
    case "GIT_PROVIDER_UNAVAILABLE":
      return "The git backend could not answer. Please try again later.";
    default:
      return "Something went wrong while loading repository files. Please try again.";
  }
}

File: cmd/api/fileshttp/files.go
package fileshttp

import (
	"context"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/lichman0405/post/cmd/api/authhttp"
	"github.com/lichman0405/post/internal/application/projects"
	"github.com/lichman0405/post/internal/gitprovider"
	"github.com/lichman0405/post/internal/observability"
)

// The Files HTTP surface (T0307): the read-only Files API over the
// platform's git backend — tree listings, file previews, commit history
// and raw downloads (docs/06: tree, preview, history, download; docs/17:
// Files is a read-only projection, never a write path).
//
// Read-only is structural here: only GET routes are registered, and the
// mux answers every other method with 405 before any handler runs
// (pinned by TestMutationMethodsRejected). The permission filter is the
// shared project read gate — the same projects.Service instance that
// serves /api/v1/projects — resolved through the guard's principal, so
// anonymous callers read public projects only (public_policy) while
// private projects stay existence-hidden behind the standard 404
// (authorized = member). No provider byte is fetched before that gate
// has passed.

// Wire error codes (docs/45): stable codes on the standard envelope.
const (
	codeDisabled            = "FILES_DISABLED"
	codeProjectNotFound     = projects.CodeProjectNotFound
	codeNotProvisioned      = "PROJECT_NOT_PROVISIONED"
	codeInvalidRef          = "FILES_INVALID_REF"
	codeInvalidSHA          = "FILES_INVALID_SHA"
	codeInvalidPath         = "FILES_INVALID_PATH"
	codeInvalidLimit        = "FILES_INVALID_LIMIT"
	codePathIsDirectory     = "FILES_PATH_IS_DIRECTORY"
	codeNotFound            = "FILES_NOT_FOUND"
	codeProviderUnavailable = "GIT_PROVIDER_UNAVAILABLE"
)

// Deps carries the collaborators the files surface needs.
type Deps struct {
	// Projects is the projects service instance that serves
	// /api/v1/projects — shared on purpose: the files of a project are
	// exactly as visible as the project itself.
	Projects *projects.Service
	// Files is the T0307 read service. Nil means the feature is disabled
	// (missing configuration): every route answers 503 naming the missing
	// keys instead of a misleading 404.
	Files *gitprovider.FilesReader
	// Missing names the configuration keys that disabled the feature
	// (key names only — never values).
	Missing []string
}

// API is the mounted files surface.
type API struct {
	handlers *handlers
}

type handlers struct {
	projects *projects.Service
	files    *gitprovider.FilesReader
	missing  []string
}

// New wires the surface.
func New(deps Deps) *API {
	return &API{handlers: &handlers{
		projects: deps.Projects,
		files:    deps.Files,
		missing:  deps.Missing,
	}}
}

// Register mounts the files routes on the guarded /api/v1 mux. Only GET
// routes exist — the Files API has no mutation methods, structurally: the
// Go 1.22+ mux answers 405 for every other method on these paths.
func (a *API) Register(mux *http.ServeMux) {
	h := a.handlers
	mux.HandleFunc("GET /api/v1/projects/{projectId}/files/tree", h.handleTree)
	mux.HandleFunc("GET /api/v1/projects/{projectId}/files/content", h.handleContent)
	mux.HandleFunc("GET /api/v1/projects/{projectId}/files/history", h.handleHistory)
	mux.HandleFunc("GET /api/v1/projects/{projectId}/files/raw", h.handleRaw)
	mux.HandleFunc("GET /api/v1/projects/{projectId}/files/diff", h.handleDiff)
}

// enabled answers 503 in place when the feature is disabled, returning
// false — the same fail-loud convention as the git-token surface.
func (h *handlers) enabled(w http.ResponseWriter, r *http.Request) bool {
	if h.files != nil {
		return true
	}
	authhttp.WriteError(w, r, http.StatusServiceUnavailable, codeDisabled,
		"files reads are disabled: missing configuration "+strings.Join(h.missing, ", "))
	return false
}

// authorize runs the permission filter: the shared project read gate.
// Anonymous callers read public projects only; private projects answer
// the existence-hiding 404 for everyone who is not a member (the same
// shape as the projects surface — read_files public_policy / authorized
// resolution, specs/policies/permissions-matrix.csv).
func (h *handlers) authorize(w http.ResponseWriter, r *http.Request, projectID string) bool {
	if _, err := h.projects.Get(r.Context(), reader(r), projectID); err != nil {
		h.projectsError(w, r, err)
		return false
	}
	return true
}

// reader mirrors projectshttp's transport-layer Reader resolution:
// a resolved session makes the caller authenticated, its absence makes
// them anonymous (reads never 401 — the service hides what it must).
func reader(r *http.Request) projects.Reader {
	p, ok := authhttp.PrincipalFrom(r.Context())
	if !ok {
		return projects.Reader{}
	}
	return projects.Reader{UserID: p.User.ID, Authenticated: true}
}

// treePayload is the tree-listing envelope. The resolved SHA lets the
// client pin follow-up reads (history) to the exact listing it saw.
type treePayload struct {
	Ref     string             `json:"ref"`
	Path    string             `json:"path"`
	SHA     string             `json:"sha"`
	Entries []treeEntryPayload `json:"entries"`
}

type treeEntryPayload struct {
	Name string `json:"name"`
	Path string `json:"path"`
	Type string `json:"type"`
	Mode string `json:"mode"`
	Size int64  `json:"size"`
	SHA  string `json:"sha"`
}

// filePayload is the single-file envelope: tree facts plus the safe
// preview. It carries nothing but the file's own git facts and (for
// pointer manifests) the pointer's own fields — private blob metadata
// never appears here (acceptance criterion, docs/17 §3).
type filePayload struct {
	Name        string              `json:"name"`
	Path        string              `json:"path"`
	Type        string              `json:"type"`
	Mode        string              `json:"mode"`
	Size        int64               `json:"size"`
	SHA         string              `json:"sha"`
	Kind        string              `json:"kind"` // text | binary | too_large | symlink | gitlink
	Content     string              `json:"content"`
	Truncated   bool                `json:"truncated"`
	Target      string              `json:"target,omitempty"`
	BlobPointer *blobPointerPayload `json:"blob_pointer,omitempty"`
}

type blobPointerPayload struct {
	ContentHash string `json:"content_hash"`
	SizeBytes   int64  `json:"size_bytes"`
	BlobID      string `json:"blob_id,omitempty"`
}

type historyPayload struct {
	Ref     string          `json:"ref"`
	Path    string          `json:"path"`
	Entries []commitPayload `json:"entries"`
}

type commitPayload struct {
	SHA         string    `json:"sha"`
	Author      string    `json:"author"`
	AuthorEmail string    `json:"author_email"`
	Date        time.Time `json:"date"`
	Message     string    `json:"message"`
}

// handleTree: GET /api/v1/projects/{projectId}/files/tree?ref=&path= —
// one directory listing at ref ("" = main, "" path = root).
func (h *handlers) handleTree(w http.ResponseWriter, r *http.Request) {
	if !h.enabled(w, r) {
		return
	}
	projectID := r.PathValue("projectId")
	if !h.authorize(w, r, projectID) {
		return
	}
	ref, path := r.URL.Query().Get("ref"), r.URL.Query().Get("path")
	listing, err := h.files.Tree(r.Context(), projectID, ref, path)
	if err != nil {
		h.filesError(w, r, err)
		return
	}
	payload := treePayload{
		Ref:     defaultedRef(ref),
		Path:    path,
		SHA:     listing.SHA,
		Entries: make([]treeEntryPayload, 0, len(listing.Entries)),
	}
	for _, e := range listing.Entries {
		payload.Entries = append(payload.Entries, treeEntryPayload{
			Name: e.Name, Path: e.Path, Type: e.Type, Mode: e.Mode, Size: e.Size, SHA: e.SHA,
		})
	}
	authhttp.WriteJSON(w, http.StatusOK, payload)
}

// handleContent: GET /api/v1/projects/{projectId}/files/content?ref=&path=
// — one file's safe preview. path is required (a directory answers 400:
// listing is the tree endpoint's job).
func (h *handlers) handleContent(w http.ResponseWriter, r *http.Request) {
	if !h.enabled(w, r) {
		return
	}
	projectID := r.PathValue("projectId")
	if !h.authorize(w, r, projectID) {
		return
	}
	ref, path := r.URL.Query().Get("ref"), r.URL.Query().Get("path")
	view, err := h.files.File(r.Context(), projectID, ref, path)
	if err != nil {
		h.filesError(w, r, err)
		return
	}
	payload := filePayload{
		Name:      view.Name,
		Path:      view.Path,
		Type:      view.Type,
		Mode:      view.Mode,
		Size:      view.Size,
		SHA:       view.SHA,
		Kind:      view.Kind,
		Content:   view.Content,
		Truncated: view.Truncated,
		Target:    view.Target,
	}
	if view.BlobPointer != nil {
		payload.BlobPointer = &blobPointerPayload{
			ContentHash: view.BlobPointer.ContentHash,
			SizeBytes:   view.BlobPointer.SizeBytes,
			BlobID:      view.BlobPointer.BlobID,
		}
	}
	authhttp.WriteJSON(w, http.StatusOK, payload)
}

// handleHistory: GET /api/v1/projects/{projectId}/files/history?ref=&path=&limit=
// — the commit history touching path at ref, newest first.
func (h *handlers) handleHistory(w http.ResponseWriter, r *http.Request) {
	if !h.enabled(w, r) {
		return
	}
	projectID := r.PathValue("projectId")
	if !h.authorize(w, r, projectID) {
		return
	}
	ref, path := r.URL.Query().Get("ref"), r.URL.Query().Get("path")
	limit := 0
	if v := r.URL.Query().Get("limit"); v != "" {
		n, err := strconv.Atoi(v)
		if err != nil || n < 1 {
			authhttp.WriteError(w, r, http.StatusBadRequest, codeInvalidLimit,
				"limit must be a positive integer")
			return
		}
		limit = n
	}
	entries, err := h.files.History(r.Context(), projectID, ref, path, limit)
	if err != nil {
		h.filesError(w, r, err)
		return
	}
	payload := historyPayload{
		Ref:     defaultedRef(ref),
		Path:    path,
		Entries: make([]commitPayload, 0, len(entries)),
	}
	for _, e := range entries {
		payload.Entries = append(payload.Entries, commitPayload{
			SHA: e.SHA, Author: e.Author, AuthorEmail: e.AuthorEmail,
			Date: e.Date, Message: e.Message,
		})
	}
	authhttp.WriteJSON(w, http.StatusOK, payload)
}

// handleRaw: GET /api/v1/projects/{projectId}/files/raw?ref=&path= — the
// download channel: provider bytes streamed straight through, with a
// browser-safe disposition. No envelope, no metadata, no execution
// context: Content-Type is octet-stream and the attachment disposition
// plus nosniff keep the payload from ever rendering in a browser context.
func (h *handlers) handleRaw(w http.ResponseWriter, r *http.Request) {
	if !h.enabled(w, r) {
		return
	}
	projectID := r.PathValue("projectId")
	if !h.authorize(w, r, projectID) {
		return
	}
	ref, path := r.URL.Query().Get("ref"), r.URL.Query().Get("path")
	raw, err := h.files.Raw(r.Context(), projectID, ref, path)
	if err != nil {
		h.filesError(w, r, err)
		return
	}
	defer raw.Body.Close()
	w.Header().Set("Content-Type", "application/octet-stream")
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename=%q`, safeFilename(path)))
	w.Header().Set("X-Content-Type-Options", "nosniff")
	if raw.Size > 0 {
		w.Header().Set("Content-Length", strconv.FormatInt(raw.Size, 10))
	}
	w.WriteHeader(http.StatusOK)
	if _, err := io.Copy(w, raw.Body); err != nil {
		observability.LoggerFromContext(r.Context()).Warn(
			"files: raw stream interrupted", "error", err.Error())
	}
}

// handleDiff: GET /api/v1/projects/{projectId}/files/diff?sha= — one
// commit's raw patch, streamed through unparsed (the raw diff view). The
// disposition stays inline so the patch renders as plain text in the
// browser, and nosniff keeps the payload from ever being sniffed into
// anything executable.
func (h *handlers) handleDiff(w http.ResponseWriter, r *http.Request) {
	if !h.enabled(w, r) {
		return
	}
	projectID := r.PathValue("projectId")
	if !h.authorize(w, r, projectID) {
		return
	}
	sha := r.URL.Query().Get("sha")
	raw, err := h.files.CommitPatch(r.Context(), projectID, sha)
	if err != nil {
		h.filesError(w, r, err)
		return
	}
	defer raw.Body.Close()
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	if raw.Size > 0 {
		w.Header().Set("Content-Length", strconv.FormatInt(raw.Size, 10))
	}
	w.WriteHeader(http.StatusOK)
	if _, err := io.Copy(w, raw.Body); err != nil {
		observability.LoggerFromContext(r.Context()).Warn(
			"files: diff stream interrupted", "error", err.Error())
	}
}

// safeFilename reduces one repository path to a Content-Disposition-safe
// basename: separators and anything outside [A-Za-z0-9._-] collapse to
// "-", and an empty result becomes "download". Path values are validated
// before this point; this is the header-injection guard, not validation.
func safeFilename(path string) string {
	base := path
	if i := strings.LastIndexByte(base, '/'); i >= 0 {
		base = base[i+1:]
	}
	var b strings.Builder
	for i := 0; i < len(base); i++ {
		c := base[i]
		if c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' ||
			c >= '0' && c <= '9' || c == '.' || c == '_' || c == '-' {
			b.WriteByte(c)
		} else {
			b.WriteByte('-')
		}
	}
	if b.Len() == 0 {
		return "download"
	}
	return b.String()
}

// defaultedRef mirrors the service's empty-ref default so the envelope
// echoes the ref actually read.
func defaultedRef(ref string) string {
	if ref == "" {
		return "main"
	}
	return ref
}

// projectsError maps projects-service errors with the same codes the
// projects surface uses (a shared service deserves a shared vocabulary).
// The denied read of a private project is the existence-hiding 404 —
// indistinguishable from an unknown project, by design (docs/45).
func (h *handlers) projectsError(w http.ResponseWriter, r *http.Request, err error) {
	switch {
	case errors.Is(err, projects.ErrProjectNotFound):
		authhttp.WriteError(w, r, http.StatusNotFound, codeProjectNotFound,
			"project not found")
	case errors.Is(err, projects.ErrForbidden):
		authhttp.WriteError(w, r, http.StatusForbidden, codeProjectNotFound,
			"you may not read this project")
	default:
		observability.LoggerFromContext(r.Context()).Error(
			"files: project lookup failed", "error", err.Error())
		authhttp.WriteError(w, r, http.StatusServiceUnavailable, "SERVICE_UNAVAILABLE",
			"the project store could not answer")
	}
}

// filesError maps service/provider errors to the wire. Provider
// credential rejections are an operator problem (the service token is
// wrong) — 503 naming it, never a 401 to the caller.
func (h *handlers) filesError(w http.ResponseWriter, r *http.Request, err error) {
	switch {
	case errors.Is(err, gitprovider.ErrInvalidRef):
		authhttp.WriteError(w, r, http.StatusBadRequest, codeInvalidRef,
			"invalid ref: use a branch name or a full commit SHA")
	case errors.Is(err, gitprovider.ErrInvalidSHA):
		authhttp.WriteError(w, r, http.StatusBadRequest, codeInvalidSHA,
			"invalid sha: use a commit SHA (7-40 hex characters)")
	case errors.Is(err, gitprovider.ErrInvalidPath):
		authhttp.WriteError(w, r, http.StatusBadRequest, codeInvalidPath, err.Error())
	case errors.Is(err, gitprovider.ErrIsDirectory):
		authhttp.WriteError(w, r, http.StatusBadRequest, codePathIsDirectory,
			"the path names a directory; list it with the tree endpoint")
	case errors.Is(err, gitprovider.ErrProjectNotFound):
		authhttp.WriteError(w, r, http.StatusNotFound, codeProjectNotFound,
			"project not found")
	case errors.Is(err, gitprovider.ErrRepoNotProvisioned):
		authhttp.WriteError(w, r, http.StatusConflict, codeNotProvisioned,
			"the project's repository is not provisioned yet — retry after provisioning completes")
	case errors.Is(err, gitprovider.ErrNotFound):
		authhttp.WriteError(w, r, http.StatusNotFound, codeNotFound,
			"no such file, directory or ref in the project's repository")
	case errors.Is(err, gitprovider.ErrUnauthorized):
		authhttp.WriteError(w, r, http.StatusServiceUnavailable, codeProviderUnavailable,
			"the git provider rejected the platform's credentials — check POST_GITEA_TOKEN")
	case errors.Is(err, context.Canceled):
		// A client that goes away mid-read is not a server error to log.
		authhttp.WriteError(w, r, http.StatusRequestTimeout, "REQUEST_CANCELED",
			"the request was canceled")
	default:
		observability.LoggerFromContext(r.Context()).Error(
			"files: read failed", "error", err.Error())
		authhttp.WriteError(w, r, http.StatusServiceUnavailable, codeProviderUnavailable,
			"the git provider could not answer")
	}
}
